src/interpreter.rs

```diff
@@ -1,222 +1,436 @@
                             }
                         }
-                        BinaryOperator::Contains => {
-                            let rhs = match &p.rhs {
-                                Value::String(s) => s,
-                                _ => unreachable!(),
-                            };
-                            let lhs = match lhs_value {
-                                Value::String(s) => s,
-                                _ => unreachable!(),
-                            };
-
-                            if lhs.contains(rhs) {
+                    }
+                    (Value::IpAddr(l), Value::IpCidr(r)) => {
+                        if r.contains(l) {
+                            matched = true;
+                            if any {
                                 return true;
                             }
                         }
                     }
+                    _ => unreachable!(),
+                },
+                BinaryOperator::NotIn => {
+                    let rhs = match &self.rhs {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
+                    let lhs = match lhs_value {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
+
+                    if !rhs.contains(lhs) {
+                        if any {
+                            return true;
+                        }
+
+                        matched = true;
+                    }
                 }
+                BinaryOperator::Contains => {
+                    let rhs = match &self.rhs {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
+                    let lhs = match lhs_value {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
 
-                // no match for all values
-                false
+                    if lhs.contains(rhs) {
+                        return true;
+                    }
+                }
+            } // match
+
+            if !any && !matched {
+                // all and nothing matched
+                return false;
             }
-        }
+        } // for iter
+
+        // if we reached here, it means that `any` did not find a match,
+        // or we passed all matches for `all`. So we simply need to return
+        // !any && lhs_values.len() > 0 to cover both cases
+        return !any && lhs_values.len() > 0;
     }
 }
+
+#[test]
+fn test_predicate() {
+    let mut mat = Match::new();
+    let mut schema = schema::Schema::default();
+    schema.add_field("my_key", Type::String);
+    let mut ctx = Context::new(&schema);
+
+    // check when value list is empty
+    // check if all values match starts_with foo -- should be false
+    let p = Predicate {
+        lhs: ast::Lhs {
```

Comment:
I noticed that the `ast::` namespace is only used in tests. In this case we can write `use crate::ast;` inside `test_predicate` instead since the above code does not use it.